attribute-projection-registry/build.gradle.kts

```diff
@@ -1,15 +1,42 @@
 plugins {
   `java-library`
   jacoco
-  id("org.hypertrace.jacoco-report-plugin")
-  id("org.hypertrace.publish-plugin")
+  alias(commonLibs.plugins.hypertrace.jacoco)
+  alias(commonLibs.plugins.hypertrace.publish)
+}
+
+configurations {
+  compileClasspath {
+    resolutionStrategy.activateDependencyLocking()
+  }
+  runtimeClasspath {
+    resolutionStrategy.activateDependencyLocking()
+  }
+}
+
+dependencyLocking {
+  lockMode.set(LockMode.STRICT)
+}
+
+tasks.register("resolveAndLockAll") {
```

Comment:
basically a utility to make it easy to generate the locks across all projects. Ref: https://docs.gradle.org/current/userguide/dependency_locking.html#ex-resolving-all-configurations

Reply:
How does this work ? What does it ensure ?

Reply:
This updates the lock files. The build will resolve all versions exactly from the lock files, meaning builds are reproducible (which they are anyways with static versions, but we're intentionally introducing dynamic versions here to support automatic upgrading).

So builds work normally, there's just the extra lock file (you'll see a couple in this PR). By running `gw resolveAndLockAll --write-locks` you can update those lock files. In a later comment - https://github.com/hypertrace/attribute-service/pull/172#discussion_r1304886875 - I discuss that I planning on adding an action that does this periodically/on trigger as well.

Reply:
Does it mean once has to run this task locally to generate the new lock file if any dep changes are made? Otherwise, gradle would continue to use the existing ones in the file?

Reply:
I'll definitely write up some documentation on this if we go forward with the locks, but in short:

- with any gradle command, you can append the `--write-locks` flag and it will update/create lock files for any dependencies it resolves.
- This task is there to make it easy to do that for all of the projects in a multi-project build at once (basically it forces resolution of all the configs). A built in way is just `gw dependencies --write-locks`, the issue with that is it works on one project at a time. Other commands, like `check` or `build` would also likely be sufficient for our case, but I figured this makes it a little more explicit and would handle projects using custom configurations that aren't resolved by those built in tasks too. Ref: https://docs.gradle.org/current/userguide/dependency_locking.html#lock_all_configurations_in_one_build_execution
- If you add or upgrade a dependency, then try to run the build without updating the lock, it will fail. If you _downgrade_ a dependency, that will not fail (basically only if conflict resolution doesn't match the lock, it's a failure).
<img width="767" alt="image" src="https://github.com/hypertrace/attribute-service/assets/45047841/e8693f2d-0da3-4954-8f30-38b2a1003e8d">.